include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h

```diff
@@ -786,6 +788,128 @@ struct __gen_transform_input
     _UnaryOp __unary_op;
 };
 
+template <typename _BinaryPred>
+struct __gen_red_by_seg_reduce_input
+{
+    // Returns the following tuple:
+    // (new_seg_mask, value)
+    // size_t new_seg_mask : 1 for a start of a new segment, 0 otherwise
+    // ValueType value     : Current element's value for reduction
+    template <typename _InRng>
+    auto
+    operator()(const _InRng& __in_rng, std::size_t __id) const
+    {
+        const auto __in_keys = std::get<0>(__in_rng.tuple());
+        const auto __in_vals = std::get<1>(__in_rng.tuple());
+        using _ValueType = oneapi::dpl::__internal::__value_t<decltype(__in_vals)>;
+        const std::size_t __new_seg_mask = __id > 0 && !__binary_pred(__in_keys[__id - 1], __in_keys[__id]);
+        return oneapi::dpl::__internal::make_tuple(__new_seg_mask, _ValueType{__in_vals[__id]});
+    }
+    _BinaryPred __binary_pred;
+};
+
+template <typename _BinaryPred>
+struct __gen_red_by_seg_scan_input
+{
+    // Returns the following tuple:
```

Comment:
In other reduce_then_scan based API implementations, we are able to reuse the `..._reduce_input` generator inside the `..._scan_input`, but I can see why we may not want to do that here to consolidate branches.  It would be nice to rely on the same code to prevent redundancy and better show connection here.  I was thinking about ways to do it (perhaps with helper functions called by the call operators of each) but it may not be worth it.

I'll let you decide on this.

Reply:
I took a look at doing this. It is possible to add some helper utilities shared between the reduce / scan input generation, but I do not think it improves the simplifies the code too much unless it is written in a way that has some more conditional checks which I want to avoid.

With the small differences in each branch in `__gen_red_by_seg_scan_input`, I think it is clearest to leave it as is.